File: day10/main.dart
import 'dart:async';
import 'dart:io';
import 'dart:convert';

Future<List<String>> getLines(filename) async {
  List<String> lines = [];
  await new File(filename)
      .openRead()
      .map(utf8.decode)
      .transform(new LineSplitter())
      .forEach((line) {
    lines.add(line);
  });
  return lines;
}

Future<int> getOpenCount(List<String> line) async {
  int count = 0;
  for (final brace in line) {
    if (brace == '(' || brace == '[' || brace == '{' || brace == '<') {
      count += 1;
    }
  }
  return count;
}

void main(List<String> arguments) async {
  String filename = 'input.txt';
  if (arguments.length == 1) {
    filename = arguments[0];
  }

  List<String> lines = await getLines(filename);

  int part1 = 0;
  int part2 = 0, score;
  Map<String, String> pairs = {')': '(', ']': '[', '}': '{', '>': '<'};
  List<String> currentOpen = [];
  List<int> scores = [];
  bool lineIsCorrupt = false;

  for (final line in lines) {
    currentOpen = [];
    lineIsCorrupt = false;
    for (final brace in line.split('')) {
      if (brace == '(' || brace == '[' || brace == '{' || brace == '<') {
        currentOpen.add(brace);
      } else {
        if (pairs[brace] == currentOpen.last) {
          currentOpen.removeLast();
        } else {
          switch (brace) {
            case ')':
              part1 += 3;
              break;
            case ']':
              part1 += 57;
              break;
            case '}':
              part1 += 1197;
              break;
            case '>':
              part1 += 25137;
              break;
          }
          if (brace == ')' || brace == ']' || brace == '}' || brace == '>') {
            lineIsCorrupt = true;
            break;
          }
        }
      }
    }
    if (!lineIsCorrupt) {
      score = 0;
      for (final brace in currentOpen.reversed.toList()) {
        switch (brace) {
          case '(':
            score = (score * 5) + 1;
            break;
          case '[':
            score = (score * 5) + 2;
            break;
          case '{':
            score = (score * 5) + 3;
            break;
          case '<':
            score = (score * 5) + 4;
            break;
        }
      }
      scores.add(score);
    }
  }

  scores.sort();
  part2 = scores[(scores.length / 2).floor()];
  print('part 1: $part1');
  print('part 2: $part2');
}

File: day10/main.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strings"
)

func main() {
	var filename string = "input.txt"
	if len(os.Args) == 2 {
		filename = os.Args[1]
	}
	input, err := os.ReadFile(filename)
	if err != nil {
		panic(err)
	}

	points1 := map[string]int{
		")": 3, "]": 57, "}": 1197, ">": 25137,
	}
	points2 := map[string]int{
		"(": 1, "[": 2, "{": 3, "<": 4,
	}

	var scores, score = []int{}, 0
	var part1, part2 = 0, 0

	lines := strings.Trim(string(input), "\n")
	for _, syntax := range strings.Split(lines, "\n") {
		for strings.Contains(syntax, "()") || strings.Contains(syntax, "[]") || strings.Contains(syntax, "{}") || strings.Contains(syntax, "<>") {
			syntax = strings.Replace(syntax, "()", "", -1)
			syntax = strings.Replace(syntax, "[]", "", -1)
			syntax = strings.Replace(syntax, "{}", "", -1)
			syntax = strings.Replace(syntax, "<>", "", -1)
		}

		if strings.Contains(syntax, ")") || strings.Contains(syntax, "]") || strings.Contains(syntax, "}") || strings.Contains(syntax, ">") {
			for _, brace := range string(syntax) {
				if val, ok := points1[string(brace)]; ok {
					part1 += val
					break
				}
			}
		} else {
			for i := len(string(syntax)) - 1; i >= 0; i-- {
				brace := string(string(syntax)[i])
				if val, ok := points2[brace]; ok {
					score = (score * 5) + val
				}
			}
			scores = append(scores, score)
			score = 0
		}
	}
	sort.Ints(scores)
	part2 = scores[len(scores)/2]
	fmt.Println("part 1:", part1, "\npart2:", part2)
}
